protocols/kad/src/query/peers/closest/disjoint.rs

```diff
@@ -0,0 +1,739 @@
+        );
+
+        let peers = known_closest_peers.into_iter().take(K_VALUE.get()).collect::<Vec<_>>();
+        let iters = (0..config.parallelism.get())
+            // NOTE: All [`ClosestPeersIter`] share the same set of peers at
+            // initialization. The [`ClosestDisjointPeersIter.contacted_peers`]
+            // mapping ensures that a successful response from a peer is only
+            // ever passed to a single [`ClosestPeersIter`]. See
+            // [`ClosestDisjointPeersIter::on_success`] for details.
+            .map(|_| ClosestPeersIter::with_config(config.clone(), target.clone(), peers.clone()))
+            .collect::<Vec<_>>();
+
+        let iters_len = iters.len();
+
+        ClosestDisjointPeersIter {
+            config,
+            iters,
+            contacted_peers: HashMap::new(),
+            // Wraps around, thus iterator 0 will be queried first.
+            last_queried: (iters_len - 1).into(),
+        }
+    }
+
+    pub fn on_failure(&mut self, peer: &PeerId) {
+        // All peer failures are reported to all queries and thus to all peer
+        // iterators. If this iterator never started a request to the given peer
+        // ignore the failure.
+        if let Some(PeerState{ initiated_by, additionally_awaited_by, response }) = self.contacted_peers.get_mut(peer) {
+            *response = ResponseState::Failed;
+
+            self.iters[*initiated_by].on_failure(peer);
+
+            for i in additionally_awaited_by {
+                self.iters[*i].on_failure(peer);
+            }
+        }
+    }
+
+    pub fn on_success<I>(&mut self, peer: &PeerId, closer_peers: I)
+    where
+        I: IntoIterator<Item = PeerId>,
+    {
+        if let Some(PeerState{ initiated_by, additionally_awaited_by, response }) = self.contacted_peers.get_mut(peer) {
+            // Mark the response as succeeded for future iterators querying this
+            // peer. There is no need to keep the `closer_peers` around, given
+            // that they are only passed to the first iterator.
+            *response = ResponseState::Succeeded;
+
+            // Pass the new `closer_peers` to the iterator that first contacted
+            // the peer.
+            self.iters[*initiated_by].on_success(peer, closer_peers);
+
+            for i in additionally_awaited_by {
+                // Only report the success to all remaining not-first iterators.
+                // Do not pass the `closer_peers` in order to uphold the
+                // S/Kademlia disjoint paths guarantee.
+                self.iters[*i].on_success(peer, std::iter::empty());
+            }
+        }
+    }
+
+    pub fn is_waiting(&self, peer: &PeerId) -> bool {
+        if let Some(PeerState{ initiated_by, additionally_awaited_by, .. }) = self.contacted_peers.get(peer) {
+            for i in std::iter::once(initiated_by).chain(additionally_awaited_by.iter()) {
+                if self.iters[*i].is_waiting(peer) {
```

Comment:
Calling this method here seems tautological, given the definition of `initiated_by` and `additionally_awaited_by`. In any case, as mentioned in the other comment, I think this method can simply call `is_waiting` on each iterator, without making use of `contacted_peers` at all.